### PR TITLE
Put Atbash first and try to animate/break up the text inside it

### PR DESCRIPTION
**Pull request by alyssamw**

### CODE REVIEW

**Comment by krashanoff** [comment]
Alyssa, I'm liking the ideas here! Only concern is that the animations will fire on every render call, so in this situation, you'll see animations repeat if you click the "brutus what's up" message to decode it. In clicking it, you reset the animation, which resets the decode animation, which leads to the user clicking on the "brutus what's up" message, which leads to...

**Comment by alyssamw** [comment]
> Alyssa, I'm liking the ideas here! Only concern is that the animations will fire on every render call, so in this situation, you'll see animations repeat if you click the "brutus what's up" message to decode it. In clicking it, you reset the animation, which resets the decode animation, which leads to the user clicking on the "brutus what's up" message, which leads to...

Leo thank u 🥺

File: src/Components/Intro/Intro.js
Comment (krashanoff) [inline src/Components/Intro/Intro.js]: I noticed you used a lot of line breaks here to get the desired look. This works visually, but can be dangerous on account of the fact that your actual anchor link getting pushed up or down the page from where it appears. Instead, try something kind of like this:

```suggestion
                    {ciphers.map((cipher, i) => {
                        return(
                            <button class="button is-static is-large is-family-secondary has-text-weight-bold">
                                <AnchorLink offset={-5} href={`#${cipher}`}>
                                    {cipherPretty[i]} Cipher
                                </AnchorLink>
                            </button>
                        )
                    })}

```

File: src/Components/Vigenere/Vigenere.js
Comment (krashanoff) [inline src/Components/Vigenere/Vigenere.js]: Might be worthwhile experimenting with Bulma's [spacers](https://bulma.io/documentation/helpers/spacing-helpers/) and/or putting them into separate paragraphs.

File: src/Components/AtbashIntro/AtbashIntro.js
Comment (krashanoff) [inline src/Components/AtbashIntro/AtbashIntro.js]: It might be a good idea to style these with the [cursor property](https://www.w3schools.com/cssref/pr_class_cursor.asp) to indicate to the user visually that these text blurbs will take action on click! I recall, though, that you were interested in triggering some animations once the user scrolls to a particular part of the page. Recall our discussion about the vertical navbar and its callback functions. Have you had a chance to play around with firing an Anime [timeline animation](https://animejs.com/documentation/#timelineBasics)?

File: src/Components/AtbashIntro/AtbashIntro.js
Comment (krashanoff) [inline src/Components/AtbashIntro/AtbashIntro.js]: There's a lot of things going on here with timing your animations. Something to chew on that might make your life easier: rather than tracking the state of your three animations in three separate arrays, and working with manual timeouts, you could instead play around with [cascading lists](https://alain.xyz/libraries/react-anime#cascading-lists) in `react-anime`.

File: src/App.js
Comment (krashanoff) [inline src/App.js]: Hmm. Looking at this, things are rather verbose. I wonder if there's something I could do to on the `<VertNav />` end to make your life easier 🤔

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Alyssa! Sorry for such the huge delay on getting this out. Things are looking good, but I have a few changes I'd like for you to make, along with a few ideas for more sweeping modifications to make life easier for you.

I also noticed that there is a merge conflict. I will resolve that once it's time to merge.

File: src/Components/Intro/Intro.js
Comment (krashanoff) [inline src/Components/Intro/Intro.js]: A good way of handling this would be:
```suggestion
                            <div className="row_space">
                                <AnchorLink offset={-5} href={"#"+cipher} className="button is-static is-large is-family-secondary has-text-weight-bold">
                                    {cipherPretty[i]} Cipher
                                </AnchorLink>
```

It's kind of cheating, since we are applying the `.button` class to a link, but this is what we did with the [links in the title](https://github.com/uclaacm/cipher-salad/blob/master/src/Components/Title/Title.js#L68).

File: src/Components/Caesar/Caesar.js
Comment (krashanoff) [inline src/Components/Caesar/Caesar.js]: I'm seeing a new bug around here. Now, when the user clicks on the message to decode it, it decodes for a split second before animating in the reverse direction.

**Review by krashanoff — COMMENTED** [review]
Hey Alyssa, thanks for starting on the changes I requested! I may update this later with more comments should they arise. For now, though, it'll just be a quick review.

**Comment by mattxwang** [comment]
Also, merge conflict? May want to merge master in before we go ahead.

**Comment by krashanoff** [comment]
> Also, merge conflict? May want to merge master in before we go ahead.

I was planning on resolving the conflict when it was time to merge, but it isn't a huge issue to merge it now.

File: src/Components/Intro/Intro.js
Comment (alyssamw) [reply src/Components/Intro/Intro.js]: Thanks Leo. In this case I took out the is-static class property so the buttons are still clickable

File: src/Components/Vigenere/Vigenere.js
Comment (alyssamw) [reply src/Components/Vigenere/Vigenere.js]: Thanks for the suggestion

File: src/Components/AtbashIntro/AtbashIntro.js
Comment (alyssamw) [reply src/Components/AtbashIntro/AtbashIntro.js]: Thanks again, was wondering how to do the clicky guy. Addressed in 52a9946

I wasn't sure if I preferred to reveal the animations via timeline or click. I did click because I feel like that would be easier to take in for kids with slower (or faster) reading speeds, so I foregoed a timed approach.

File: src/Components/Caesar/Caesar.js
Comment (alyssamw) [reply src/Components/Caesar/Caesar.js]: Reverse direction? I see it the same direction each time--but yes it does seem less intuitive now to have Brutus pop up for only a second after each click. Will work on clicking decodes one way each time.
So click -> encode
click again -> decode
click again -> encode

File: src/Components/AtbashIntro/AtbashIntro.js
Comment (alyssamw) [reply src/Components/AtbashIntro/AtbashIntro.js]: I was totally fixated on cascading lists as timer based, so thanks for making me take another look. Gave it a different whirl in 53a95d1. Let me know what you think.

File: src/Components/Caesar/Caesar.js
Comment (krashanoff) [inline src/Components/Caesar/Caesar.js]: Another way of doing things, just if you'd like to remove a member function:

```suggestion
                        {ciphertext.map(c => <span>{c}</span>)}

```

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Alyssa, things are coming together. Cascading lists are looking good, as are the toggling animations of the encrypted message to Brutus. Great work!

This will likely be the last series of changes - sorry for being a stickler!

Right now, the encrypted text for "click me!" is the other way around - it is encrypted until you click it. Was this intended?

three encrypted messages

Cursor interactions on the "can you crack these messages?" is still set to the default. This can be fixed with setting the `cursor` property:

can you crack these messages?

File: src/Components/Caesar/Caesar.js
Comment (alyssamw) [reply src/Components/Caesar/Caesar.js]: I'll keep Lisha's code here, because her putCharsInSpans function I believe creates the nice gradual effect of each letter being replaced-- straight mapping them replaces them almost all at once and doesn't look as ~special~ imo.

**Review by krashanoff — COMMENTED** [review]
Hey Alyssa! Last review. Promise.

I'm seeing that when we reach the portion that is telling kids to click on our messages, the motion of the two are joined, so when one is clicked, they both bounce in from left once more:

image

This could be fixed by placing each into their own animation container. As for the vertical nav business, I think we'll have to leave that to its own PR. Thanks so much for sticking it out through these changes! Appreciate it.

**Comment by alyssamw** [comment]
> Hey Alyssa! Last review. Promise.
> ...
> Thanks so much for sticking it out through these changes! Appreciate it.

Thank you for all the feedback over the course of my many commits! Haha

File: src/Components/main.css
Comment (krashanoff) [inline src/Components/main.css]: Just a general note, but you can use shorthand CSS here to set `30px` on all sides.

```suggestion
    margin: 30px;
```

File: src/Components/Vigenere/Vigenere.js
Comment (krashanoff) [inline src/Components/Vigenere/Vigenere.js]: In place of `div`s with the `.subtitle` class, I'd recommend using `p` tags (with the `.subtitle` class).

File: src/Components/AtbashIntro/AtbashIntro.js
Comment (krashanoff) [inline src/Components/AtbashIntro/AtbashIntro.js]: A more descriptive name here would be preferred. Totally optional, though. `onClick` is a super common handler name.

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]: We should generally avoid using `<br>` tags, but when we do, we can just use `<br>` to keep it tidy. Bear in mind, the reason we avoid using `<br>` tags is because in semantic HTML [they represent a carriage return, or break in phrase](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/br). Instead, consider using one of the [spacing helpers in Bulma](https://bulma.io/documentation/helpers/spacing-helpers/).

File: src/App.js
Comment (krashanoff) [inline src/App.js]: This is definitely a rough one since the sections of the app are compartmentalized. This will be another refactor down the line. For now, a good fix to the dangling `'game-link'` would be to change the ID of the component in question to, say, `gameLink`.

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]:
```suggestion
```

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js:51]: Any call of this function like:

```js
<div>
    {putCharsInSpans('some text')}
</div>
```

...can be replaced with:

```js
<div>
    {Array.from('some text').map((c, i) => <span key={i}>{c}</span>}
</div>
```

Just a trade-off here, but saves a little complexity without sacrificing readability.

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]: This can be boiled down significantly:

```suggestion
  decryptHelper = (msgIdx, substrIdx) => {
    let original = this.state.messages[msgIdx];
    original = atbashEncode( original.substring(0, substrIdx+1)) + original.substring(substrIdx + 1);
    this.setState({
      messages: Array.from(this.state.messages).splice(msgIdx, 1, original),
    });
  }
```

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]:
```suggestion
    this.timeouts = [];
```

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]: You can simplify this down and boost performance if you swap out your array of state variables for a single one. I think you'd be surprised with what you can do with `map` as opposed to `splice`. We can discuss this one a little more in Slack since there's a lot to chew on here.

File: src/Components/Title/Recap.js
Comment (krashanoff) [inline src/Components/Title/Recap.js]: If `this.animate()` is only called once, it might be worth removing the member function and replacing it with an arrow one.

File: src/Components/Atbash/AtbashInput.js
Comment (krashanoff) [inline src/Components/Atbash/AtbashInput.js]: The way things are set up right now, your timeouts roll over, rather than being cleared. Assuming an `Array`, we can fix this:

```suggestion
  componentWillUnmount() {
    this.timeouts.forEach(t => clearTimeout(t));
  }
```

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey again Alyssa, things are coming along! Good lord this review inflated. C'est la broad-scope PR. Should have caught the encroaching scope of this all earlier - my mistake.

While the core functionality is there (and looking great!), there are a lot of minor adjustments to make. I have left a bunch of comments of spots where things should be cleaned up a bit, along with some suggestions. Let me know if there are any questions at all!

While it's completely outside the scope of this PR, I'd like for us to use [husky and lint-staged](https://codeburst.io/continuous-integration-lint-staged-husky-pre-commit-hook-test-setup-47f8172924fc?gi=77c6c0abbe83) to keep the code consistent. I'll likely do a pass-over after this is merged.

**Comment by mattxwang** [comment]
Wow, this PR is becoming unwieldy!

@alyssamw you may find it helpful to batch-commit from the files tab for most of the suggestions. Tysm for working on this PR for the past few weeks, I know it's been gruelling but it really adds to the quality of the app ❤️

File: src/Components/Atbash/AtbashInput.js
Comment (alyssamw) [reply src/Components/Atbash/AtbashInput.js:51]: Just a note that putCharsInSpans does a little bit more than call map on the string array.

It actually uses Array.prototype.map, and then calls .call() on it, to create the nice gradual effect Lisha animated. 

To fit in one line it will look like

`{Array.prototype.map.call(ciphertext, c => (c === " ") ? <span>&nbsp;</span> : <span key={c}>{c}</span>)}`

Which works too, but I'm not sure if that's as readable than the helper function.

File: src/Components/Atbash/AtbashInput.js
Comment (alyssamw) [reply src/Components/Atbash/AtbashInput.js]: Mind if I take the time to flesh that out this weekend?

**Review by krashanoff — APPROVED** [review]
Looking great Alyssa! Just looked through the changes and things are looking excellent.

Apologies for being such a stickler this PR, and for being so particular about the code. Things are looking good to merge, thanks for sticking with it for all this time - feature burnout, as Matt mentioned, is very real, and is something I'd like to help avoid inducing.

In any case, great work once again. Take it easy! 😄

**Pull request merged**
